File: HashTable/350_IntersectionOfTwoArraysII.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-07-02 10:47:23


class Solution(object):
    def intersect(self, nums1, nums2):
        nums1_dict = {}
        for n in nums1:
            nums1_dict[n] = nums1_dict.get(n, 0) + 1
        ans = []
        for n in nums2:
            if nums1_dict.get(n, 0) > 0:
                ans.append(n)
                nums1_dict[n] = nums1_dict.get(n, 0) - 1
        return ans


class Solution_2(object):
    def intersect(self, nums1, nums2):
        nums1.sort()
        nums2.sort()
        i, j = 0, 0
        ans = []
        while i < len(nums1) and j < len(nums2):
            if nums1[i] == nums2[j]:
                ans.append(nums1[i])
                i += 1
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                j += 1
        return ans

"""
[]
[]
[1,2,2,2,3]
[3,2,2,2,3]
[1, 2, 2, 1]
[2, 2]
"""

File: HashTable/350_IntersectionOfTwoArraysII.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-02 10:59:01
 */

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        map<int, int> nums1_hash;
        for(auto &n : nums1){
            nums1_hash[n]++;
        }
        vector<int> ans;
        for(auto &n : nums2){
            if(nums1_hash.find(n) != nums1_hash.end() && nums1_hash[n]-- > 0){
                ans.push_back(n);
            }
        }
        return ans;
    }
};


class Solution_2{
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        int i=0, j=0;
        vector<int> ans;
        while(i<nums1.size() && j<nums2.size()){
            if(nums1[i] == nums2[j]){
                ans.push_back(nums1[i]);
                i++;
                j++;
            }
            else if(nums1[i] > nums2[j]){
                j++;
            }
            else{
                i++;
            }
        }
        return ans;
    }
};

/*
[]
[]
[1,2,2,2,3]
[3,2,2,2,3]
[1, 2, 2, 1]
[2, 2]
*/
